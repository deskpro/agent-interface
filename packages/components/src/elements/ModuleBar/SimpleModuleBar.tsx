import * as React from "react";

import ModuleBar from "./ModuleBar";
import ModuleBarMenu from "./ModuleBarMenu";

export interface SimpleModuleBarItem {
  icon: string;
  label: string;
  name: React.Key;
}

export interface SimpleModuleBarPortalLink {
  icon: "plane" | "plane-violet" | "plane-orange";
  title: string;
  subtitle: string;
  link: string;
}

export type SimpleModuleBarProps = {
  className?: string;
  items: Array<SimpleModuleBarItem | "-">;
  portalMenuTitle?: string;
  portalMenuLinks?: SimpleModuleBarPortalLink[];
  onItemClick: (itemName: React.Key, e: React.SyntheticEvent) => void;
};

const SimpleModuleBar = ({
  className,
  items,
  onItemClick,
  portalMenuTitle = "Portals",
  portalMenuLinks = []
}) => {
  const [activeItem, setActiveItem] = React.useState<React.Key>();
  const [isPortalsMenuVisible, setMenuPortalsVisibility] = React.useState<
    boolean
  >(false);

  return (
    <ModuleBar className={className}>
      {items.map((item: SimpleModuleBarItem | "-", index: number) => {
        if (item === "-") {
          // eslint-disable-next-line react/no-array-index-key
          return <ModuleBar.Separator key={`separator_${index}`} />;
        }
        const { name, ...itemProps } = item;
        return (
          <ModuleBar.Item
            {...itemProps}
            key={name}
            isActive={activeItem === name}
            onClick={e => {
              setActiveItem(name);
              onItemClick(name, e);
            }}
          />
        );
      })}
      {!!portalMenuLinks.length && (
        <ModuleBar.Separator key="portal-separator" />
      )}
      {!!portalMenuLinks.length && (
        <ModuleBar.Item
          icon="portal"
          key="portal"
          onClick={() => setMenuPortalsVisibility(!isPortalsMenuVisible)}
        >
          <ModuleBarMenu
            title={portalMenuTitle}
            isVisible={isPortalsMenuVisible}
          >
            {portalMenuLinks.map((link: SimpleModuleBarPortalLink) => (
              <ModuleBarMenu.Item {...link} key={link.title} />
            ))}
          </ModuleBarMenu>
        </ModuleBar.Item>
      )}
    </ModuleBar>
  );
};

export default SimpleModuleBar;
