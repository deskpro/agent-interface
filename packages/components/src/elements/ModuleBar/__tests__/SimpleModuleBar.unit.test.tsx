import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import SimpleModuleBar from "../SimpleModuleBar";

const modules = [
  { icon: "ticket", name: "ticket", label: "Tickets" },
  { icon: "crm", name: "crm", label: "CRM" },
  { icon: "chat", name: "chat", label: "Chats" },
  { icon: "feedback", name: "feedback", label: "Feedback" },
  { icon: "publish", name: "publish", label: "Publish" },
  { icon: "tasks-sidebar", name: "tasks-sidebar", label: "Tasks" },
  "-",
  { icon: "reports", name: "reports", label: "Reports" },
  { icon: "admin", name: "admin", label: "Admin" },
  { icon: "billing", name: "billing", label: "Billing" }
];

const portalLinks = [
  {
    icon: "plane",
    title: "Raven Paper",
    subtitle: "support.ravenpaper.com",
    link: "http://support.ravenpaper.com"
  },
  {
    icon: "plane-orange",
    title: "Raven Paper Crafts",
    subtitle: "support.ravenpapercrafts.com",
    link: "http://support.ravenpapercrafts.com"
  },
  {
    icon: "plane-violet",
    title: "RP Europe",
    subtitle: "support.rpeurope.com",
    link: "http://support.rpeurope.com"
  }
];

describe("<SimpleModuleBar />", () => {
  it("should match its snapshot", () => {
    const { container } = render(
      <SimpleModuleBar
        className="module-bar"
        items={modules}
        portalMenuLinks={portalLinks}
        onItemClick={jest.fn()}
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render separator", () => {
    const { container } = render(
      <SimpleModuleBar
        className="module-bar"
        items={modules}
        portalMenuLinks={portalLinks}
        onItemClick={jest.fn()}
      />
    );
    expect(
      container.querySelector(".dp-ModuleBar-separator")
    ).toBeInTheDocument();
  });

  it("should toggle active items on click", () => {
    const clickHandler = jest.fn();
    const { getByText, container } = render(
      <SimpleModuleBar
        className="module-bar"
        items={modules}
        portalMenuLinks={portalLinks}
        onItemClick={clickHandler}
      />
    );

    expect(getByText("CRM")).toBeInTheDocument();
    const ticketItem = container.querySelector(".Icon--ticket");
    fireEvent.click(ticketItem);
    expect(
      ticketItem.parentElement.classList.contains("is-active")
    ).toBeTruthy();
    expect(clickHandler).toBeCalledTimes(1);
    expect(clickHandler.mock.calls[0][0]).toBe("ticket");
  });

  it("should open the portal menu on click", () => {
    const clickHandler = jest.fn();
    const { container } = render(
      <SimpleModuleBar
        className="module-bar"
        items={modules}
        portalMenuLinks={portalLinks}
        onItemClick={clickHandler}
      />
    );

    const portalItem = container.querySelector(".Icon--portal");
    fireEvent.click(portalItem);
    expect(
      container
        .querySelector(".dp-ModuleBarMenu")
        .classList.contains("is-visible")
    ).toBeTruthy();
  });
});
