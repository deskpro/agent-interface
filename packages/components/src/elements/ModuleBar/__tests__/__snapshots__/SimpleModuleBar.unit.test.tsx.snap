// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SimpleModuleBar /> should match its snapshot 1`] = `
<div
  class="dp-ModulebarContainer module-bar"
>
  <ul
    class="dp-ModuleBar"
  >
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--ticket"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        Tickets
      </span>
    </li>
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--crm"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        CRM
      </span>
    </li>
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--chat"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        Chats
      </span>
    </li>
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--feedback"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        Feedback
      </span>
    </li>
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--publish"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        Publish
      </span>
    </li>
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--tasks-sidebar"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        Tasks
      </span>
    </li>
    <li
      class="dp-ModuleBar-separator"
    />
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--reports"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        Reports
      </span>
    </li>
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--admin"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        Admin
      </span>
    </li>
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--billing"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <span
        class="dp-IconTooltip"
      >
        Billing
      </span>
    </li>
    <li
      class="dp-ModuleBar-separator"
    />
    <li
      class="dp-ModuleBarList dp-ModuleBar-item"
    >
      <span
        class="dp-Icon Icon--portal"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="21"
          width="21"
        />
      </span>
      <div
        class="dp-ModuleBarMenu"
      >
        <span
          class="dp-ModuleBarMenu-title"
        >
          Portals
        </span>
        <ul>
          <li
            class="dp-ModuleBarMenu-item"
          >
            <a
              href="http://support.ravenpaper.com"
            >
              <span
                class="dp-Icon Icon--plane"
              />
              <span>
                <span
                  class="dp-ModuleBarMenu-itemTitle"
                >
                  Raven Paper
                  ,
                </span>
                <span
                  class="dp-ModuleBarMenu-itemSubtitle"
                >
                  support.ravenpaper.com
                </span>
              </span>
            </a>
          </li>
          <li
            class="dp-ModuleBarMenu-item"
          >
            <a
              href="http://support.ravenpapercrafts.com"
            >
              <span
                class="dp-Icon Icon--plane-orange"
              />
              <span>
                <span
                  class="dp-ModuleBarMenu-itemTitle"
                >
                  Raven Paper Crafts
                  ,
                </span>
                <span
                  class="dp-ModuleBarMenu-itemSubtitle"
                >
                  support.ravenpapercrafts.com
                </span>
              </span>
            </a>
          </li>
          <li
            class="dp-ModuleBarMenu-item"
          >
            <a
              href="http://support.rpeurope.com"
            >
              <span
                class="dp-Icon Icon--plane-violet"
              />
              <span>
                <span
                  class="dp-ModuleBarMenu-itemTitle"
                >
                  RP Europe
                  ,
                </span>
                <span
                  class="dp-ModuleBarMenu-itemSubtitle"
                >
                  support.rpeurope.com
                </span>
              </span>
            </a>
          </li>
        </ul>
      </div>
    </li>
  </ul>
</div>
`;
